src/Illuminate/Routing/RouteBinding.php

```diff
@@ -67,7 +70,23 @@ public static function forModel($container, $class, $callback = null)
             // throw a not found exception otherwise we will return the instance.
             $instance = $container->make($class);
 
-            if ($model = $instance->resolveRouteBinding($value)) {
+            $parent = $route->parentOfParameter($key);
+
+            if ($parent instanceof UrlRoutable && in_array($key, array_keys($route->bindingFields()))) {
+                try {
+                    $model = $parent->resolveChildRouteBinding(
+                        $key, $value, $field
+                    );
+                } catch (BadMethodCallException $exception) {
+                    $model = $parent->resolveChildRouteBinding(
```

Comment:
This is using for the undefined explicit model binding name.

For example:
```php
Route::model('special_post', Post::class);

Route::get('foo/{user}/{special_post:slug}');
```

At first, it looks for the `specialPosts` relationship in the `user` model, If `specialPosts` does not exist then it's gonna use `posts` relationship.

If you think it's a Unnecessary behavior let me know.